src/Collision.cc

```diff
@@ -256,23 +262,59 @@ void Collision::CalculateInertial(
   gz::math::Inertiald &_inertial,
   const ParserConfig &_config)
 {
-  // Check if density was not set during load & send a warning
-  // about the default value being used
-  if (!this->dataPtr->densitySetAtLoad)
+  this->CalculateInertial(
+    _errors, _inertial, _config, std::nullopt, ElementPtr());
+}
+
+/////////////////////////////////////////////////
+void Collision::CalculateInertial(
+  sdf::Errors &_errors,
+  gz::math::Inertiald &_inertial,
+  const ParserConfig &_config,
+  const std::optional<double> &_density,
+  sdf::ElementPtr _autoInertiaParams)
+{
+  // Order of precedence for density:
+  double density;
```

Comment:
This logic made me realize that the link could just call `SetDensity` and `SetAutoInertiaParams` on the collision object before calling `CalculateInertial` and we wouldn't need a new overload. I like this because it's more "functional", but not sure if it's worth the overload.

Reply:
If the Collision density / `auto_inertia_params` values are unset and we call `Collision::SetDensity` / `Collision::SetAutoInertiaParams`, then those values will no longer be unset. This might or might not be important. It could matter to the `ToElement` functions, but I realized while writing this that I didn't update `Link::ToElement` or `Collision::ToElement` in this pull request.

Currently `Link::ToElement` does not export a `//inertial/@auto` attribute and explicitly exports all inertia values. `Collision::ToElement` unconditionally exports a `//collision/density` element whether the value was set or unset in the original SDFormat model.

I like the current, functional approach but am not sure what we expect from the output of the `ToElement` functions.

Reply:
> If the Collision density / auto_inertia_params values are unset and we call Collision::SetDensity / Collision::SetAutoInertiaParams, then those values will no longer be unset.

Good point. We won't have a way of knowing if those parameters were set by the user in the `<collision>` or not. So, let's go with the current approach.

>  but I realized while writing this that I didn't update Link::ToElement or Collision::ToElement in this pull request.

I didn't remember that either. In general, how do we remember to add things to `ToElement` when we add something to the spec? I wonder if we can write a test (not in this PR, of course) that calls `ToElement` and checks the output against the spec.

 > Currently Link::ToElement does not export a //inertial/@auto attribute and explicitly exports all inertia values. Collision::ToElement unconditionally exports a //collision/density element whether the value was set or unset in the original SDFormat model.

We have a general problem with values not set by the user showing up in `ToElement` (https://github.com/gazebosim/sdformat/issues/867). In most of our DOM objects, we don't keep track of whether the element/attribute was set by the user, so it's hard to figure out what to do in `ToElement`. Now that you've made `Collision::density` optional, I think we can actually call `Element::SetExplicitlySetInFile` on the element so it's printed properly according to the print config (https://github.com/gazebosim/sdformat/blob/b41b48b612db47a30409b80b4d1bb793ab734377/src/Element.cc#L592)

Regarding `//inertial/@auto`, the behavior was supposed to be determined by [ConfigureResolveAutoInertials](https://github.com/gazebosim/sdformat/blob/b41b48b612db47a30409b80b4d1bb793ab734377/include/sdf/ParserConfig.hh#L55), but I don't think it's working correctly because it's exporting all inertia values unconditionally. 

I think the logic should be the following in pseudocode

```
auto==true && SAVE_CALCULATION:
  export all inertia
auto==true && not SAVE_CALCULATION:
 export `//inertial/auto`
```

But it looks like we need more enum values for "skip calculation during load and do not save calculation" and "calculate during load but do not save calculation".

Reply:
I've updated `Collision::ToElement` in https://github.com/gazebosim/sdformat/pull/1335/commits/8d91b6a962a90670bfafb64b41acb37e54c2040c and `Link::ToElement` in https://github.com/gazebosim/sdformat/pull/1335/commits/9460341a6eb3aa01b2ef023e0c3599bbb6e46b47

> We have a general problem with values not set by the user showing up in `ToElement` (#867). In most of our DOM objects, we don't keep track of whether the element/attribute was set by the user, so it's hard to figure out what to do in `ToElement`. Now that you've made `Collision::density` optional, I think we can actually call `Element::SetExplicitlySetInFile` on the element so it's printed properly according to the print config

I didn't use `Element::SetExplicitlySetInFile`; please advise further if it's necessary

Reply:
> Regarding `//inertial/@auto`, the behavior was supposed to be determined by [ConfigureResolveAutoInertials](https://github.com/gazebosim/sdformat/blob/b41b48b612db47a30409b80b4d1bb793ab734377/include/sdf/ParserConfig.hh#L55), but I don't think it's working correctly because it's exporting all inertia values unconditionally.
> 
> I think the logic should be the following in pseudocode
> 
> ```
> auto==true && SAVE_CALCULATION:
>   export all inertia
> auto==true && not SAVE_CALCULATION:
>  export `//inertial/auto`
> ```
> 
> But it looks like we need more enum values for "skip calculation during load and do not save calculation" and "calculate during load but do not save calculation".

Can we address this in a follow-up ticket?

Reply:
> I didn't use Element::SetExplicitlySetInFile; please advise further if it's necessary

I removed `<density>` from `link_4`  in `test/sdf/inertial_stats_auto.sdf` and ran `gz sdf -p` and it didn't print the default density, so I think we're good.

> Can we address this in a follow-up ticket?

sure!

Reply:
> > Can we address this in a follow-up ticket?
> 
> sure!

https://github.com/gazebosim/sdformat/issues/1371